Content.Client/Chemistry/UI/ChemMasterWindow.xaml.cs

```diff
@@ -102,36 +138,57 @@ private ReagentButton MakeReagentButton(string text, ChemMasterReagentAmount amo
+            _currentState = castState;
+
             if (castState.UpdateLabel)
                 LabelLine = GenerateLabel(castState);
-            UpdatePanelInfo(castState);
-
-            var output = castState.OutputContainerInfo;
 
+            // Ensure the Panel Info is updated, including UI elements for Buffer Volume, Output Container and so on
+            UpdatePanelInfo(castState);
+    
             BufferCurrentVolume.Text = $" {castState.BufferCurrentVolume?.Int() ?? 0}u";
-
+    
             InputEjectButton.Disabled = castState.InputContainerInfo is null;
-            OutputEjectButton.Disabled = output is null;
-            CreateBottleButton.Disabled = output?.Reagents == null;
-            CreatePillButton.Disabled = output?.Entities == null;
-
+            OutputEjectButton.Disabled = castState.OutputContainerInfo is null;
+            CreateBottleButton.Disabled = castState.OutputContainerInfo?.Reagents == null;
+            CreatePillButton.Disabled = castState.OutputContainerInfo?.Entities == null;
+            
+            UpdateDosageFields(castState);
+        }
+        
+        //assign default values for pill and bottle fields.
+        private void UpdateDosageFields(ChemMasterBoundUserInterfaceState castState)
+        {
+            var output = castState.OutputContainerInfo;
             var remainingCapacity = output is null ? 0 : (output.MaxVolume - output.CurrentVolume).Int();
             var holdsReagents = output?.Reagents != null;
             var pillNumberMax = holdsReagents ? 0 : remainingCapacity;
             var bottleAmountMax = holdsReagents ? remainingCapacity : 0;
+            var bufferVolume = castState.BufferCurrentVolume?.Int() ?? 0;
 
-            PillTypeButtons[castState.SelectedPillType].Pressed = true;
-            PillNumber.IsValid = x => x >= 0 && x <= pillNumberMax;
-            PillDosage.IsValid = x => x > 0 && x <= castState.PillDosageLimit;
-            BottleDosage.IsValid = x => x >= 0 && x <= bottleAmountMax;
+            PillDosage.Value = (int)Math.Min(bufferVolume, castState.PillDosageLimit);
 
-            if (PillNumber.Value > pillNumberMax)
-                PillNumber.Value = pillNumberMax;
-            if (BottleDosage.Value > bottleAmountMax)
-                BottleDosage.Value = bottleAmountMax;
-        }
+            // Avoid division by zero
+            if (PillDosage.Value > 0)
+            {
+                PillNumber.Value = Math.Min(bufferVolume / PillDosage.Value, pillNumberMax);
+            }
+            else
+            {
+                PillNumber.Value = 0;
+            }
 
+            BottleDosage.Value = Math.Min(bottleAmountMax, bufferVolume);
+        }
+        //forces the UI to refresh using cached data on swapping tab
+        private void OnTabChanged(int tabIndex)
+        {
+            if (_currentState != null)
+            {
+                UpdatePanelInfo(_currentState);
+            }
+        }
```

Comment:
Shouldn't all the tabs just be updated anyway?

Reply:
I am uncertain i follow your meaning

Reply:
If I remember it right, the tab contents exist at all times, even if you’ve switched away from the tab in question. So the entire UI can be updated when a new state gets received.

Reply:
If I remove the code you indicate, this is the behavior I encounter. 

Uploading 2024-11-25 15-23-04.mp4…

Reply:
If this code is not implemented i get this behaviour.

https://github.com/user-attachments/assets/7899e1f3-30d4-4486-8807-a161c004153d

If you know why this is and can offer a way to avoid it I'll happily work with that.

Reply:
You need to debug and find why it's not being called on state changes.

Reply:
found the problem, is good now.